service levels: update `client_state` when SL is altered

Jadw1 opened the pull request:
When any changes were made to service levels or roles, the client had to restart connection for the changes to take effect.

This PR makes the `client_state` params are updated when any service level is altered or dropped, by registering a subscriber on service level configuration change.
Update of `client_state` params isn't instant because service level configuration is updated in 10s interval.

This PR doesn't fix the need of restarting the connection when 
- a service level is attached to a role
- a service level is detached from a role
- one role is granted to another

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/495/

eliransin commented on service/qos/service_level_controller.cc:
```diff
@@ -219,16 +220,22 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
 
 future<>  service_level_controller::notify_service_level_added(sstring name, service_level sl_data) {
```
I think this should be refactored differently, we can enhance the subscriber api by adding "before" and "after" version. That way it will be more verbose and you will not need to supply one more key when subscribing.
for example, on_service_level_add, will become two functions on_before_service_level_add and on_after_service_level_add

avikivity replied:
Agree.

There's the question of exception safety (preexisting). If one of the notifiers throws, how do we handle it?

avikivity replied:
I guess we could add a cancel notification. "before" would be allowed to throw, but cancel and after not. "before" would be required to reserve all necessary resources.

eliransin reviewed (changes requested):
The approach (of registering the client state) is good IMO, makes the core of the code
very simple. The juggling with the before and after subscribers should be refactored in a way that will allow all combinations. I left a comment with suggestion about how to do it.
It will also make it more simple for the code reader since he will not have to "remember" if a subscriber registered with a before or after key.
Hope it makes sense :slightly_smiling_face:

avikivity commented on service/client_state.hh:
```diff
@@ -126,6 +128,18 @@ private:
 
     workload_type _workload_type = workload_type::unspecified;
 
+    class qos_configuration_change_client_state_subscriber : public qos::qos_configuration_change_subscriber {
```
You could make client_state inherit privately from configuration_change_subscriber.

But if you're using indirection, let's squeeze some compile time decoupling and forward-declare the class here, and define it in .cc, so you don't have to add a #include to .hh.

Jadw1 replied:
Removed the subscriber class and made client state inherit privately from configuration_change_subscriber

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1290/

Jadw1 commented:
@eliransin ping

eliransin commented on test/boost/service_level_controller_test.cc:
```diff
@@ -61,6 +65,10 @@ struct qos_configuration_change_suscriber_simple : public qos_configuration_chan
+    }
 };
 
 std::ostream& operator<<(std::ostream& os, const add_op& op) {
```
I would expect the methods will do something to validate that the apis are called at the right moment (for example validating that while running in an "on_before" method the service level is still unchanged / added /removed and in the after the expected change has happened.

Jadw1 replied:
Honestly I don't quite understand why and how to do it? 
Do you mean to for instance add assert(if service level is in configuration) before this line
https://github.com/scylladb/scylladb/blob/master/service/qos/service_level_controller.cc#L223

Jadw1 replied:
Added

mykaul commented:
@Jadw1 - please resolve review comments and lets get it reviewed and done with - this PR is being worked on for months - lets drive it to completion ASAP.

Jadw1 commented:
> @Jadw1 - please resolve review comments and lets get it reviewed and done with - this PR is being worked on for months - lets drive it to completion ASAP.

The PR is ready to be merged. I've reminded @eliransin today and now I've rebased on latest master because there was a merge conflict.
The only thing is I don't get the point of latest @eliransin comment.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4794/
- Duration: 2 hr 21 min
- Builder: spider6.cloudius-systems.com

avikivity commented on service/client_state.hh:
```diff
@@ -347,6 +347,18 @@ public:
     future<> has_function_access(const sstring& ks, const sstring& function_signature, auth::permission p) const;
+    future<> has_functions_access(data_dictionary::database db, auth::permission p) const;
+    future<> has_functions_access(data_dictionary::database db, const sstring& ks, auth::permission p) const;
+    future<> has_function_access(data_dictionary::database db, const sstring& ks, const sstring& function_signature, auth::permission p) const;
```
What are these?

Jadw1 replied:
it was a leftover after merge I didn't notice. removed

avikivity commented:
Please add a test.

Jadw1 commented:
> Please add a test.

Done. I had to do it as a boost test because I couldn't find a way to check if the change(timeout and workload_type was updated) actually happened from pytest perspective

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4870/
- Duration: 2 hr 25 min
- Builder: spider6.cloudius-systems.com

Jadw1 commented:
> Please add a test.

Changelog:
- added test for client_state paramenters update
- updated existing test to include new methods of `qos_configuration_change_subscriber` (@eliransin's comment I didn't get at first)

@avikivity @eliransin

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:x: - Sanity Tests
:white_check_mark: - Unit Tests
#### Failed Tests (1/21912):
- [test_can_limit_compaction_throughput](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4964/testReport/junit/throughput_limits_test/TestCompactionLimitThroughput/Tests___Sanity_Tests___test_can_limit_compaction_throughput) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_can_limit_compaction_throughput)
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4964/
- Duration: 2 hr 13 min
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4983/
- Duration: 6 hr 50 min
- Builder: monster

nyh commented:
@eliransin I see you were in the middle of a review, can you please take a look and make a decision (approve / request changes)? I don't want to merge it in the middle of your review.

eliransin reviewed (approved):
LGTM
I saw the tests now take into account the new operations.

eliransin commented:
@nyh finished reviewing and approved :slightly_smiling_face:

nyh commented:
> When any changes were made to service levels or roles, the client had to restart connection for the changes to take effect.
> 
> This PR makes the `client_state` params are updated when any service level is altered or dropped, by registering a subscriber on service level configuration change. Update of `client_state` params isn't instant because service level configuration is updated in 10s interval.
> 
> This PR doesn't fix the need of restarting the connection when
> 
>     * a service level is attached to a role
> 
>     * a service level is detached from a role
> 
>     * one role is granted to another

Why doesn't the PR do this - is it just because you plan to do this as followups, or because it's not important (?) or because it's impossible to do?
If you don't plan to this immediately as a followup, should you perhaps open an issue about these missing cases?

nyh reviewed (commented):
I left some small comments about typos, but I have a bigger question:

I didn't understand why we needed to complicate everything with all these notifiers, lifetime issues, and so on. If I understand correctly, we don't support service-level changes in the middle of the request, so all of this is only relevant between different requests. So I have an idea and want to run it by you:

What if the client_state just saved the authenticated user's **name**, and no pointer to no service-level object?
Before each an every request, we would fetch the object using this name. This will allow not only to change the service level,  but also to completely delete it (what, if I understood correctly, your patch doesn't fix).

avikivity commented on service/qos/qos_configuration_change_subscriber.hh:
```diff
@@ -16,14 +16,19 @@ namespace qos {
         virtual future<> on_before_service_level_add(service_level_options slo, service_level_info sl_info) = 0;
-        /** This callback is going to be called just after the service level is removed **/
+        /** This callback is going to be called just after the service level is available **/
+        virtual future<> on_after_service_level_add(service_level_options slo, service_level_info sl_info) = 0;
```
noexcept?

avikivity replied:
Or rather, a statement about how this may not fail, since an exception can be returned in the future too.

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -240,10 +246,17 @@ future<> service_level_controller::notify_service_level_updated(sstring name, se
             sl_it->second.slo = slo;
+            _subscribers.thread_for_each([name, slo_before, slo] (qos_configuration_change_subscriber* subscriber) {
+                try {
+                    subscriber->on_after_service_level_change(slo_before, slo, {name}).get();
+                } catch (...) {
```
Why do we need a thread here?

I guess you just copied the other function. Please follow up later with de-threading.

avikivity commented on service/client_state.cc:
```diff
@@ -271,3 +271,34 @@ future<> service::client_state::maybe_update_per_service_level_params() {
+future<> service::client_state::on_before_service_level_change(qos::service_level_options slo_before, qos::service_level_options slo_after, qos::service_level_info sl_info) {
+    return make_ready_future<>();
+}
+future<> service::client_state::on_after_service_level_change(qos::service_level_options slo_before, qos::service_level_options slo_after, qos::service_level_info sl_info) {
```
Please add blank lines between functions.

avikivity commented on test/boost/service_level_controller_test.cc:
```diff
@@ -106,13 +139,66 @@ SEASTAR_THREAD_TEST_CASE(subscriber_simple) {
+            });
+        });
+    });
+}
```
Better as a coroutine.

Jadw1 commented:
Closing without merging because we are solving this with raft service levels https://github.com/scylladb/scylladb/pull/19085
